File: Reqnroll/Chapter14/14.5-DataTableAccessorPattern/Before/WIMP.Specs/StepDefinitions/MenuAdminStepDefinitions.cs
using Reqnroll;

using WIMP.App.Services;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class MenuAdminStepDefinitions(MenuService menuService)
{
    [Given("the restaurant menu is")]
    public void GivenTheRestaurantMenuIs(DataTable menuItemsTable)
    {
        foreach (var row in menuItemsTable.Rows)
        {
            string name = row["name"];
            decimal price = menuItemsTable.ContainsColumn("price")
                ? decimal.Parse(row["price"])
                : DomainDefaults.MenuItemPrice;
            int calories = menuItemsTable.ContainsColumn("calories")
                ? int.Parse(row["calories"])
                : DomainDefaults.MenuItemCalories;
            string ingredients = menuItemsTable.ContainsColumn("ingredients")
                ? row["ingredients"]
                : DomainDefaults.MenuItemIngredients;

            menuService.AddPizzaItem(name, price, calories, ingredients);
        }
    }
}

File: CucumberJVM/Chapter14/14.5-DataTableAccessorPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/MenuAdminStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.services.MenuService;
import com.wimp.app.specs.support.DomainDefaults;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.*;

import java.math.BigDecimal;

public class MenuAdminStepDefinitions {

    private final MenuService menuService;

    public MenuAdminStepDefinitions(MenuService menuService) {
        this.menuService = menuService;
    }

    @Given("the restaurant menu is")
    public void theRestaurantMenuIs(DataTable menuItemsTable) {
        for (var row : menuItemsTable.asMaps()) {
            var name = row.get("name");
            var price = row.containsKey("price")
                ? new BigDecimal(row.get("price"))
                : DomainDefaults.PIZZA_PRICE;
            var calories = row.containsKey("calories")
                ? Integer.parseInt(row.get("calories"))
                : DomainDefaults.PIZZA_CALORIES;
            var ingredients = row.containsKey("ingredients")
                ? row.get("ingredients")
                : DomainDefaults.PIZZA_INGREDIENTS;

            menuService.addMenuItem(name, price, calories, ingredients);
        }
    }
}
